modules/dataframe/src/io/csvreader.cpp

```diff
@@ -33,262 +33,519 @@
+        setUnitRegexp(*unitRegexp);
+        return true;
     } else if (auto* doublePrecision = std::any_cast<bool>(&value);
                doublePrecision && key == "DoublePrecision") {
         setEnableDoublePrecision(*doublePrecision);
         return true;
+    } else if (auto* rows = std::any_cast<size_t>(&value); rows && key == "NumberOfExampleRows") {
+        setNumberOfExampleRows(*rows);
+        return true;
+    } else if (auto* locale = std::any_cast<std::string>(&value); locale && key == "Locale") {
+        setLocale(*locale);
+        return true;
+    } else if (auto* emptyField = std::any_cast<EmptyField>(&value);
+               emptyField && key == "HandleEmptyFields") {
+        setHandleEmptyFields(*emptyField);
+        return true;
     }
+
     return false;
 }
 
 std::any CSVReader::getOption(std::string_view key) {
     if (key == "Delimiters") {
         return getDelimiters();
+    } else if (key == "StripQuotes") {
+        return getStripQuotes();
     } else if (key == "FirstRowHeader") {
         return hasFirstRowHeader();
+    } else if (key == "UnitsInHeaders") {
+        return hasUnitsInHeaders();
+    } else if (key == "UnitRegexp") {
+        return getUnitRegexp();
     } else if (key == "DoublePrecision") {
         return hasDoublePrecision();
+    } else if (key == "NumberOfExampleRows") {
+        return getNumberOfExamplesRows();
+    } else if (key == "Locale") {
+        return getLocale();
+    } else if (key == "HandleEmptyFields") {
+        return getHandleEmptyFields();
     }
     return std::any{};
 }
 
-namespace detail {
+std::shared_ptr<DataFrame> CSVReader::readData(std::string_view fileName) {
+    auto file = open(fileName);
 
-bool isLineBreak(const char ch, std::istream& stream) {
-    if (ch == '\r') {
-        // consume potential LF (\n) following CR (\r)
-        if (stream.peek() == '\n') {
-            stream.get();
-        }
-        return true;
-    } else {
-        return (ch == '\n');
+    file.seekg(0, std::ios::end);
+    std::streampos len = file.tellg();
+    file.seekg(0, std::ios::beg);
+
+    if (len == std::streampos(0)) {
+        throw DataReaderException(IVW_CONTEXT, "Emtpy file: {}", fileName);
     }
+
+    return readData(file);
 }
 
-}  // namespace detail
+namespace util {
 
-std::shared_ptr<DataFrame> CSVReader::readData(std::istream& stream) const {
-    // Skip BOM if it exists. Added by for example Excel when saving csv files.
-    filesystem::skipByteOrderMark(stream);
+template <typename T>
+constexpr bool alwaysFalse() {
+    return false;
+}
 
-    if (stream.bad() || stream.fail()) {
-        throw CSVDataReaderException("Input stream in a bad state", IVW_CONTEXT);
-    }
+template <typename Func>
+size_t parse(std::string_view str, std::string_view delimiters, std::optional<size_t> expectedParts,
+             std::optional<size_t> refLine, Func&& func) {
+
+    std::string lookFor{delimiters};
+    lookFor.push_back('"');
```

Comment:
Suggest to include an `empty` varable and use it in the code below to make it more explicit
```suggestion
    const auto empty = '"';
    lookFor.push_back(empty); 
```

Reply:
it not empty, its a quote, added a comment.